src/experimental/java/org/quiltmc/loader/impl/VersionConstraintImpl.java

```diff
@@ -60,4 +82,40 @@ public boolean equals(Object o) {
 	public int hashCode() {
 		return Objects.hash(this.version, this.type);
 	}
+
+	@Override
+	public boolean matches(Version version) {
+		if (type == Type.ANY) {
+			return true;
+		}
+
+		if (version.isSemantic()) {
+			Version.Semantic semantic = version.semantic();
+			switch (type) {
+				case EQUALS:
+					return semanticVersion.compareTo(semantic) == 0;
+				case GREATER_THAN:
+					return semanticVersion.compareTo(semantic) > 0;
+				case GREATER_THAN_OR_EQUAL:
+					return semanticVersion.compareTo(semantic) >= 0;
+				case LESSER_THAN:
+					return semanticVersion.compareTo(semantic) < 0;
+				case LESSER_THAN_OR_EQUAL:
+					return semanticVersion.compareTo(semantic) <= 0;
+				case SAME_MAJOR:
+					return semanticVersion.major() == semantic.major();
+				case SAME_MAJOR_AND_MINOR:
+					return semanticVersion.major() == semantic.major()
+						&& semanticVersion.minor() == semantic.minor();
+				default:
+					throw new IllegalStateException("Unknown VersionConstraint.Type " + type);
+			}
+		} else {
+			if (version.raw().equals("${version}")) {
+				// Special cased by QMJ
+				return true;
+			}
+			return false;
```

Comment:
This returns false for a.matches(a)

Reply:
How?

`VersionConstraintImpl` doesn't implement `Version`, so it's not possible to pass the constaint back into itself (`VersionConstraintImpl.matches(Version)`).

(Although I do think it's a bit off that we can't handle string versions here, however that's what the QMJ spec suggests?)